lnst/Recipes/ENRT/ConfigMixins/DevFlowsPinningHWConfigMixin.py

```diff
@@ -0,0 +1,236 @@
+        https://man7.org/linux/man-pages/man8/ethtool.8.html
+
+        Can be "partially" used by pinning just selected flows and keeping
+        the rest unpinned - so they'll be steered by rx hash algorithm just
+        by returning `None` in :meth:`.steer_flow_by` method for unpinned flow.
+
+        This mixin does NOT pin IRQs to
+        :attr:`lnst.RecipeCommon.Perf.Measurements.BaseFlowMeasurement.Flow.receiver_cpupin`
+        CPUs, which is required for optimal performance.
+    """
+
+    def hw_config(self, config: EnrtConfiguration):
+        """
+        Pins flows generated by
+        :meth:`BaseFlowMeasurementGenerator.generate_flow_combinations`
+        to CPU queues by using `ethtool -N ... flow-type...` command.
+
+        Internal configuration is stored in
+        :attr:`EnrtConfiguration.hw_config.flow_steering_rules` attribute.
+        """
+        super().hw_config(config)
+
+        config.hw_config["flow_steering_rules"] = {}
+
+        for combination in self.generate_flow_combinations(config):
+            for flow in combination:
+                if self.steer_flow_by(flow) is None:
+                    continue  # skip unpinned flow
+
+                for steering_nic in self.steer_flow_at(flow):
+                    if steering_nic not in config.hw_config["flow_steering_rules"]:
+                        config.hw_config["flow_steering_rules"][steering_nic] = []
+
+                    config.hw_config["flow_steering_rules"][steering_nic].append(
+                        self._pin_flow(flow, steering_nic)
+                    )
+
+        return config
+
+    def _pin_flow(self, flow: Flow, steering_nic: Device) -> tuple[str, str]:
+        """
+        Method pins :param:`flow` to :any:`Flow.receiver_cpupin`.
+        Pinning is done by `ethtool -N ... flow-type ... action`,
+        not by `queue`. So other actions (by <0 action code) can
+        be specified as well.
+        See https://man7.org/linux/man-pages/man8/ethtool.8.html
+
+        :param flow: flow object to pin
+        :param steering_nic: NIC where the flow should be pinned
+
+        :returns: tuple of rule ID (for deconfig) and description of pinning.
+            That can't be dynamically generated in :meth:`.describe_hw_config`
+            as it woudln't know which flow belongs to which rule ID.
+        """
+        cpupin = flow.receiver_cpupin
+        if len(cpupin) > 1:
+            logging.warning(
+                "Flow is pinned to multiple CPUs.\
+            Why? Usual rx-hash algos will throw packets from \
+            the same flow to the same queue anyway (based on its' 5-tuple)."
+            )
+
+            logging.warning(f"Using only the first CPU: {cpupin[0]}")
+
+        steer_by = self.steer_flow_by(flow)
+        flow_type = self._get_flow_type(flow, steer_by)
+        steering_key = self._get_steering_key(flow, steer_by)
+
+        job = steering_nic.netns.run(
```

Comment:
I'm quite sure you can omit the `steering_nic` parameter as it can be fetched from `Flow.receiver_nic`. Yes, it is optional field of the `Flow` but for ENRT recipes it is always present.

To be 100% sure this works as expected, I'd just add a check if the receiver_nic is not None, if it is just skip that flow instance.

Reply:
Ack, resolving.